src/gt4py/storage/cartesian/utils.py

```diff
@@ -308,17 +308,21 @@ def allocate_gpu(
     return buffer.buffer, buffer_ndarray
 
 
-if CUPY_DEVICE == core_defs.DeviceType.ROCM:
+if CUPY_DEVICE == core_defs.DeviceType.CUDA:
+    allocate_gpu = _allocate_gpu
+elif CUPY_DEVICE == core_defs.DeviceType.ROCM:
 
-    @functools.wraps(allocate_gpu)
-    def allocate_gpu_rocm(
+    @functools.wraps(_allocate_gpu)
```

Comment:
I think we could also move here from above the definition of the ndarray subclass adding the CUDA and HIP array interfaces to make the code more readable. Something like:

```suggestion
    class CUDAArrayInterfaceNDArray(cp.ndarray):
        def __new__(cls, input_array: "cp.ndarray") -> CUDAArrayInterfaceNDArray:
            return (
                input_array
                if isinstance(input_array, CUDAArrayInterfaceNDArray)
                else cp.asarray(input_array).view(cls)
            )

        @property
        def __cuda_array_interface__(self) -> dict:
            return {
                "shape": self.shape,
                "typestr": self.dtype.descr[0][1],
                "descr": self.dtype.descr,
                "stream": 1,
                "version": 3,
                "strides": self.strides,
                "data": (self.data.ptr, False),
            }
            
        __hip_array_interface__ = __cuda_array_interface__

    @functools.wraps(_allocate_gpu)
```